Divide project into library and binary parts

tuguzT opened the pull request:
Fixes #105

tuguzT committed:
Divide project into library and binary parts

tuguzT committed:
Add `clap` feature for `cargo-gpu` library

tuguzT committed:
Add `watch` feature for `cargo-gpu` library

tombh commented:
So this allows using `cargo-gpu` as a library? I wonder if there is simple e2e test for that? And/or perhaps an example in the README of how to do that?

schell commented on Cargo.toml:
```diff
@@ -1,17 +1,25 @@
-	# uses a toolchain whose Cargo version doesn't recognise version 4 of `Cargo.lock`.
-	"crates/shader-crate-template"
+	# uses a toolchain whose Cargo version doesn't recognize version 4 of `Cargo.lock`.
+	"crates/shader-crate-template",
```
Just a nit pick here that "recognise" is British English while "recognize" is American English, and both are valid. I'm an American living in New Zealand, so I switch between the two depending on which device I'm on (due to spell check on that device).

tuguzT replied:
Oh, I see. I could revert that change, because in this case I was blindly following the suggestion of my IDE.

schell replied:
It's not necessary, this is fine.

schell commented on crates/cargo-gpu/src/build.rs:
```diff
@@ -89,6 +101,7 @@ impl Build {
             std::env::current_dir()?.display()
         );
 
+        #[cfg(feature = "watch")]
```
Something like: 

```rust
#[cfg(feature = "watch")]
let watching = self.build.watch;
#[cfg(not(feature = "watch"))]
let watching = false;

if watching {
...
} else {
...
}
```

tuguzT replied:
The main difficulty here is not the flag itself, but the code inside of `is watching` block, which is valid only if `watch` feature of `spirv-builder` is enabled.
I really want to move this logic inside of private method to not to clutter `run` method with conditional compilation handling.

schell commented on crates/cargo-gpu/src/build.rs:
```diff
@@ -109,10 +122,12 @@ impl Build {
     }
 
     /// Parses compilation result from `SpirvBuilder` and writes it out to a file
+    #[cfg(feature = "watch")]
```
I don't think we want this cfg here, because we still want to parse compilation results if we're not watching.

tuguzT replied:
I placed this just because this was marked by linter as unused (because I've cfg-featured the code block above).
Rewriting an `if` above with removing this `cfg` will resolve it.

schell commented on crates/cargo-gpu/src/lib.rs:
```diff
@@ -140,17 +144,18 @@ impl Command {
-                    command.run()?;
                 }
+
+                command.run()?;
```
Good simplification 👍

schell reviewed (changes requested):
Just needs a few tweaks to the cfg stuff. Good work!

nazar-pc commented:
> So this allows using `cargo-gpu` as a library? I wonder if there is simple e2e test for that? And/or perhaps an example in the README of how to do that?

It is already possible on `main`. This PR just allows you to no compile binary-only dependencies (like clap) when you only use it as a library, see linked issue.

tuguzT committed:
Fix incorrect `watch` feature handling in build

tuguzT commented on crates/cargo-gpu/src/build.rs:
```diff
@@ -89,6 +94,23 @@ impl Build {
+        };
+
+        #[cfg(feature = "watch")]
         if self.build.watch {
```
This logic also could be moved inside a new private method (for example, `watch()`), which could panic with `unreachable!()` if reached **without** `watch` feature. Also `build` closure could further be transformed into a method.
But is it worth the effort?

tuguzT replied:
> But is it worth the effort?

I've decided it really is...

tuguzT commented on crates/cargo-gpu/src/build.rs:
```diff
@@ -89,30 +94,52 @@ impl Build {
+                        accept.submit(parse_result);
                     }
                 })?
                 .context("unreachable")??;
```
I have a strong feeling that this line means `watch` function should return `!` or `anyhow::Result<!>`, but I'm not sure which one was in mind when this code were written.

schell replied:
Without looking at the code I'm thinking it makes most sense to use `anyhow::Result<!>`. Conceptually, watching is a never-ending loop, and there _could_ be errors along the way.

schell replied:
Ah, I see, that's on `SpirvBuilder`...

tombh replied:
Even though I originally added the watch code in `cargo-gpu` I don't remember all of the failure cases I'm afraid. I do personally use `--watch` often and it works well (just a small annoyance that it compiles twice on first run). So all I'd want to see is that the compilation logs still display and that `CTRL+C` exits.

schell replied:
I think in practice the `loop` above should never actually loop, as `thread::park` _should_ never return. But the `loop`gives us the return type we want.

I haven't actually written any of this code I'm suggesting though, so the real compiler (as opposed to my head compiler) might have something else to say about it ;).

tuguzT commented:
Also I'm not sure about APIs of library and binary parts:
- should `Cli` & `Command` structs & some binary-only features (`Config` & dump usage) be moved inside of binary part,
- what the minimal API of library part should be (for it to be usable inside of a build script)?

tuguzT commented:
I'll try to reorder commits of this branch via rebase now to de-clutter merge history.

tuguzT commented:
> > So this allows using `cargo-gpu` as a library? I wonder if there is simple e2e test for that? And/or perhaps an example in the README of how to do that?
> 
> It is already possible on `main`. This PR just allows you to not compile binary-only dependencies (like clap) when you only use it as a library, see linked issue.

I think it would be nice to add `build.rs` script for `shader-crate-template` showing the usage of `cargo-gpu` as a library, but should this be done here or in a separate PR?

schell commented on crates/cargo-gpu/Cargo.toml:
```diff
@@ -28,7 +25,11 @@ semver.workspace = true
+
+[features]
+watch = ["spirv-builder/watch"]
+clap = ["dep:clap", "spirv-builder/clap"]
```
Yes, I agree a little note would be good.

schell reviewed (changes requested):
So far so good, we just need to address the feature comment and the `watch` situation. Thank you!

Firestar99 reviewed (changes requested):
One huge red flag: We *cannot* rename the cli tool. It must stay `cargo-gpu`, otherwise you'd have to call it with `cargo gpu-cli` instead of just `cargo gpu` (iirc). Especially since we want to (more strongly) mirror cargo commands in the future, like `cargo gpu check` or `cargo gpu clippy`.

Otherwise, I strongly support the split up. I just haven't gotten around to it, or figured out a good alternative name for the "cargo-gpu-lib". Some ideas, feel free to suggest more:
* rustc_codegen_spirv_cache
* cargo_gpu_cache
* rust_gpu_cache

nazar-pc commented:
> One huge red flag: We cannot rename the cli tool.

The binary is still `cargo-gpu` (see `[[bin]]` section), only the crate name has changed.

schell commented:
> Also I'm not sure about APIs of library and binary parts:
> 
>     * Should `Cli`, `Command` structs and some binary-only features (for example, `Config` and dump usage) be moved inside of binary part?
> 
>     * What should the minimal API of library part be (for it to be usable inside of a build script)?
>       Is it just all the pub contents of `install` module & `spirv-builder` re-export (as it is currently), or is there something else what could be useful inside of a build script?
> 
> 
> This could be resolved in some separate PR, but I think this needs to be addressed at least.

I would just make most of it `pub` and call it a day. We can always go back and clean up the API if it's too complex.

tuguzT commented:
Unfortunately, I don't have much time on this very week to fix those issues, but I'll work on them ASAP.

schell commented:
So long as `cargo-gpu` works for the binary name, and you said it did, then `cargo-gpu-cli` is fine for the crate name. It's a nice and descriptive name, and folks won't really interact with it anyway.

Thanks for your great work @tuguzT!

tuguzT commented:
> Some ideas, feel free to suggest more:
> 
> * rustc_codegen_spirv_cache
> * cargo_gpu_cache
> * rust_gpu_cache

I think `rustc_codegen_spirv_cache` or `spirv-builder-cache` might suite quite well.

The former suits because, as stated in README,
> `cargo gpu` compiles a custom [codegen backend](https://doc.rust-lang.org/beta/unstable-book/compiler-flags/codegen-backend.html) for `rustc`

The latter - because `cargo-gpu` directly re-exports it and enhances its capabilities by managing its codegen.

Firestar99 commented:
Have you changed something CI related that lints is failing, or is it just failing in general? (I'll go fix it if it's on us)

tuguzT commented:
I've mostly done with more aggressive separation of library & crate parts.
I could add more error types for `cargo-gpu` via `thiserror`, but this is not the main point of the PR.

Here is the recap:
- added new crate `rustc_codegen_spirv-cache` which responsibility is to install backend codegen dylib (compile it & move inside of cache directory)
- added new crate `cargo-gpu-test-utils` which shares code for testing
  (used only in `dev-dependencies` by the crate above & `cargo-gpu` CLI part)
- `cargo-gpu-build` handles building of shader crates, this crate could be used inside of build scripts, something like this:
```rust
let compile_result = build_shader_crate(
    "../path/to/shader_crate",
    "spirv-unknown-vulkan1.2",
    SpirvBuilder::default()
        .shader_crate_default_features(false)
        .shader_crate_features(["nightly".into()])
        .spirv_metadata(SpirvMetadata::Full),
    InstallParams::default(),
)?;
```
- `cargo-gpu` is a CLI, some of its internals was just moved into the crates above

---

After review I could fix Git history by squashing commits.

But for me it feels wrong to just squash these commits for a single reason: many files are marked as new ones, but in reality these are just modified and moved into different directories. This is problematic because file history (and history of lines changed) is completely screwed by now.

While googling I came up with an idea to how to resolve it:
- firstly, create commit that just moves files inside of new directories (I could just use `anyhow` for error handling due to the fact that creating new error types pretty much exclusively contributes to file changes)
- secondly, make changes to these moved files (and replace `anyhow` with errors created with `thiserror`)
